rfd/0-job-worker.md

```diff
@@ -0,0 +1,229 @@
+
+##### Job Status
+
+Job status is extremely simple and only returns a status code and the exit_code. The status code is one of:
+
+- running: the job has not yet completed
+- completed: the job completed successfully
+- error: the job completed with a non-zero exit code
+
+##### Job Output Storage
+
+Output for each job will be maintained, in memory, as long as the worker implementation exists. This poses a potential memory leak, but for the purposes of this challenge, is an acceptable solution. Later designs would likely need to write job output to disk or possibly even remote storage.
+
+##### Job Output Streaming
+
+The worker library streams job output through an `io.ReadCloser`. The server will indicate the end of the stream, if the job has completed, with the `io.EOF` error. The client may, at any time, close the stream to indicate that it is disconnecting. The client must always close the stream when it no longer needs it to prevent memory leaks.
+
+##### Identifiers
+
+JobID is an opaque identifier that is internally mapped to the pid of the process in the host namespace. Since pid is not guaranteed to be unique, nothing internal to the library will be keyed off of the pid. The pid is still required, though, in order to signal the process to stop, for example. The `go.jetify.com/typeid` library is used to generate job ids.
+
+UserID can be any unique value as far as the library is concerned. See the section on Authorization below for more information.
+
+#### gRPC API
+
+The gRPC API follows the worker library fairly closely. It does not require the user id in the messages because user id is the client certificate's serial number, not explicitly provided. Additionally, the stream output format is slightly different since gRPC can't stream bytes, rather, it will stream one newline separated string at a time.
+
+### Security Considerations
+
+#### TLS
+
+The following TLS configuration will be used, where `clientCAs` includes the ca cert that should sign all client certs and `crt` contains the server's cert and private key. Note that cipher suites are not used because they are [not configurable](https://go.dev/blog/tls-cipher-suites) in Go when using TLS 1.3. This configuration requires and validates client certificates for all connections.
+
+```go
+tls.Config{
+    ClientAuth:   tls.RequireAndVerifyClientCert,
+    ClientCAs:    clientCAs,
+    Certificates: []tls.Certificate{crt},
+    MinVersion:   tls.VersionTLS13,
+}
+```
+
+#### Authorization
+
+The serial number of the client certificate is used as a unique user identifier and is implicitly required for all requests. It is used for authorization such that only the user that starts a job is permitted to stop it, get its status or output. This way the client doesn't have to explicitly provide any other kind of identification. The serial number is guaranteed to be unique for all certificates signed by a given authority thus making this a valid method to uniquely identify users.
```

Comment:
Is there another way where we do not need to use the certificate serial number? 
If the client certificate needs to be re-issued because it's about to expire, will we need to keep the same serial number? Are there other fields we can use instead of relying on a field that's expected to be unique per certificate?

Reply:
yeah, good point, we want to identify the user, not the cert. how about the cert subject?